Bluetooth: Audio: Add PACS for broadcast sink

Thalley opened the pull request:
Add support for using PACS and/or capabilities for
the broadcast sink role. PACS and capabilities were
originally only supported for the unicast server
role, so the PACS callbacks were moved their own
struct, as if a device supports both the unicast
server role and the broadcast sink role, it will
only have a single PACS instance.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

fixes https://github.com/zephyrproject-rtos/zephyr/issues/43788

Thalley committed:
Bluetooth: Audio: Add PACS for broadcast sink

Add support for using PACS and/or capabilities for
the broadcast sink role. PACS and capabilities were
originally only supported for the unicast server
role, so the PACS callbacks were moved their own
struct, as if a device supports both the unicast
server role and the broadcast sink role, it will
only have a single PACS instance.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented on subsys/bluetooth/audio/Kconfig.baps:
```diff
@@ -152,6 +152,8 @@ config BT_AUDIO_BROADCAST_SINK
 	bool "Bluetooth Broadcast Sink Audio Support [EXPERIMENTAL]"
 	select EXPERIMENTAL
 	select BT_ISO_SYNC_RECEIVER
+	select BT_PERIPHERAL
```
To use PACS the broadcast sink must somehow be connectable

Thalley commented on include/zephyr/bluetooth/audio/capabilities.h:
```diff
@@ -234,10 +234,12 @@ struct bt_audio_capability {
 	uint16_t context;
 	/** Capability codec reference */
 	struct bt_codec *codec;
+#if defined(CONFIG_BT_AUDIO_UNICAST_SERVER)
 	/** Capability preferences */
 	struct bt_audio_capability_pref pref;
 	/** Capability operations reference */
 	struct bt_audio_capability_ops *ops;
+#endif /* CONFIG_BT_AUDIO_UNICAST_SERVER */
 	sys_snode_t node;
```
I wonder if `pref` should be moved to a unicast server-only struct since it has no value for the broadcast sink. 

Somewhat the same with the `ops` as those are only used for the unicast server, but the broadcast sink could be refactored to use some of them (except that the broadcast sink always works on a group, and not individual streams).

Casper-Bonde-Bose replied:
Each BIS is an individual stream I guess and an BASS-client can dynamically instruct a receiver to sync to one or more BISes - hence the application needs to be able to handle individual BIS LC3 start/stop + cleanup.
Don't think it needs to be part of this PR though, unless it is easier? (Not sure how it works today from an API perspective - maybe this is already supported)

Thalley replied:
These are the `struct bt_audio_capability_ops`, which are used by the unicast server for stream configuration. 

The `struct bt_audio_stream_ops` still has the `started` and `stopped` callbacks for the broadcast sink role.

Thalley committed:
Bluetooth: Audio: Fix some bad guards in audio shell

Some functions are only used for the unicast server
but was guarded by CONFIG_BT_AUDIO_UNICAST instead of
CONFIG_BT_AUDIO_UNICAST_SERVER.

Some parts of the unicast-server-only code also had
client specific code that would never be called.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Disable HAS when unicast_server is disabled

The HAS depends on the unicast server, and should be
disabled in the `no_unicast_server` as well as depend
on the CONFIG_BT_AUDIO_UNICAST_SERVER config.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

asbjornsabo reviewed (dismissed):
LGTM

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
Rebased to fix test failure

MariuszSkamra reviewed (dismissed):
LGTM

Thalley commented:
Rebased to solve merge conflicts

Thalley committed:
Bluetooth: Audio: Add guards for calling bt_audio_capabilities

Add guards for calling bt_audio_capabilities when
CONFIG_BT_AUDIO_CAPABILITY=n

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: PACS: Fix support context if SRC/SNK disabled

If CONFIG_BT_PAC_SNK or CONFIG_BT_PAC_SRC was not enabled, the
function would not build due to missing Kconfig
(CONFIG_BT_PACS_SNK_CONTEXT or CONFIG_BT_PACS_SRC_CONTEXT).

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Casper-Bonde-Bose commented on subsys/bluetooth/shell/audio.c:
```diff
@@ -1352,17 +1318,21 @@ static int cmd_init(const struct shell *sh, size_t argc, char *argv[])
+	if (IS_ENABLED(CONFIG_BT_AUDIO_CAPABILITY)) {
+		/* Mark all supported contexts as available */
+		bt_audio_capability_set_available_contexts(BT_AUDIO_DIR_SINK,
+							   BT_AUDIO_CONTEXT_TYPE_UNSPECIFIED);
```
Maybe allow all types? If a client have "Media" content and you only allow "Unspecified" it will not be able to setup a stream.

Thalley replied:
Agreed; updated

Casper-Bonde-Bose commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/broadcast_sink_test.c:
```diff
@@ -19,6 +20,9 @@ CREATE_FLAG(pa_sync_lost);
 static struct bt_audio_broadcast_sink *g_sink;
 
+/* Mandatory support preset by both source and sink */
+static struct bt_audio_lc3_preset preset_16_2_1 = BT_AUDIO_LC3_BROADCAST_PRESET_16_2_1;
```
In general(all sample apps) I think this variable name should be reduced to `preset` or `selected_preset` such that if I want to try for example 48 khz I simply select another preset rather than having to rename the variable all the places it is used :-)

Thalley replied:
This is a test case though, but I understand what you mean. For this test case it can also easily just be called `preset`. Updated

Casper-Bonde-Bose reviewed (dismissed):
LGTM
A single suggested rename - optionally.
The suggested addition for Broadcast behavior could go into a separate PR.

Thalley committed:
Bluetooth: Audio: Change context type for BAP shell

The shell module will now use BT_AUDIO_CONTEXT_TYPE_ANY
instead of BT_AUDIO_CONTEXT_TYPE_UNSPECIFIED as that provides
a more flexible solution.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

fredrikdanebjer commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -201,17 +201,17 @@ static ssize_t supported_context_read(struct bt_conn *conn,
-	} else {
-		context.snk = BT_AUDIO_CONTEXT_TYPE_PROHIBITED;
-	}
+#if defined(CONFIG_BT_PAC_SNK)
```
Just wondering what prompted you to change from the IS_ENABLED to #if defined here

Thalley replied:
The Kconfig `CONFIG_BT_PACS_SNK_CONTEXT` is guarded by `CONFIG_BT_PAC_SNK`:
```c
if BT_PAC_SRC
config BT_PACS_SRC_CONTEXT
	hex "Supported Source Contexts"
	default 0x03ff
	range 0x0000 0x03ff
```

So we won't be able to use it unless `CONFIG_BT_PAC_SNK` is enabled (`IS_ENABLED` does not work for guarded Kconfigs)

Thalley commented:
Fixed small issue with `context.snk = sys_cpu_to_le16(CONFIG_BT_PACS_SRC_CONTEXT);` (should be `context.src`)

asbjornsabo commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1271,45 +1278,61 @@ void bt_audio_stream_cb_register(struct bt_audio_stream *stream,
  */
 
-/** @brief Register unicast server callbacks.
+/** @brief Register Publish Audio Capabilities Service callbacks.
```
The name of the service is "Published Audio Capabilities Service" - an additional "ed".
(Nitpicking, I know.)

asbjornsabo commented on include/zephyr/bluetooth/audio/audio.h:
```diff
@@ -1271,45 +1278,61 @@ void bt_audio_stream_cb_register(struct bt_audio_stream *stream,
  * @return 0 in case of success or negative value in case of error.
  */
-int bt_audio_unicast_server_location_changed(enum bt_audio_dir dir);
+int bt_audio_pacs_location_changed(enum bt_audio_dir dir);
```
Maybe outside of this PR, so possibly make a note of it.
Does this call notify a change of the locations, or just of the direction of the location?  In the latter case, the name is probably confusing.

asbjornsabo commented on subsys/bluetooth/audio/pacs_internal.h:
```diff
@@ -7,7 +7,7 @@
  */
 
-#include <zephyr/bluetooth/audio/audio.h>
+#include <bluetooth/audio/audio.h>
```
The "zephyr" part of the path should still be there.  (It has probably been the victim of rebasing across that change.)

asbjornsabo replied:
"zephyr" part of path missing

asbjornsabo reviewed (changes requested):
Fix include paths and a couple of other minor comments

carlescufi commented:
@MariuszSkamra can you please take a look?
